src/Client.php

```diff
@@ -542,21 +559,67 @@ protected function send(array $data, array $tags = [])
      */
     protected function sendMessages(array $messages)
     {
+        if (is_null($this->socket)) {
+            $this->socket = $this->connect();
+        }
+        $this->message = implode("\n", $messages);
+        $this->write();
+
+        return $this;
+    }
+
+    /**
+     * Attempt to write the current message to the socket
+     *
+     * @return bool
+     */
+    private function write()
+    {
+        if (!is_null($this->socket)) {
+            if (@fwrite($this->socket, $this->message) === false) {
+                // attempt to re-send on socket resource failure
+                $this->socket = $this->connect();
+                if (!is_null($this->socket)) {
+                    return (@fwrite($this->socket, $this->message) !== false);
+                }
+            } else {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    /**
+     * Creates a persistent connection to the udp host:port
+     *
+     * @return resource
+     * @throws ConnectionException If there is a connection problem with the host
+     */
+    protected function connect()
+    {
         $socket = @fsockopen('udp://' . $this->host, $this->port, $errno, $errstr, $this->timeout);
-        if (!$socket) {
+        if ($socket === false) {
+            $socket = null;
```

Comment:
In this way, if DataDog will be down, then common timeout for DataDog connection will be `$this->timeout * count of messages` - and this is completely wrong. Nobody wants to get down just because DataDog is down. I was in this situation on some highload project, on production, and it was a really big fail.